src/library/dao/playlistdao.cpp

```diff
@@ -980,18 +981,19 @@ void PlaylistDAO::shuffleTracks(const int playlistId, const QList<int>& position
         trackPositionIds.insert(trackBPosition, trackAId);
         newPositions.swap(newPositions.indexOf(trackAPosition),
                           newPositions.indexOf(trackBPosition));
-        QString swapQuery = "UPDATE PlaylistTracks SET position=%1 "
-                "WHERE position=%2 AND playlist_id=%3";
-        query.exec(swapQuery.arg(QString::number(-1),
-                                 QString::number(trackAPosition),
-                                 QString::number(playlistId)));
-        query.exec(swapQuery.arg(QString::number(trackAPosition),
-                                 QString::number(trackBPosition),
-                                 QString::number(playlistId)));
-        query.exec(swapQuery.arg(QString::number(trackBPosition),
-                                 QString::number(-1),
-                                 QString::number(playlistId)));
 
+        qSwap(tracksPositionRowIds[trackAPosition - 1].position,
+                  tracksPositionRowIds[trackBPosition - 1].position);
```

Comment:
I think we should go the save route and not assume Position = i + 1, It might be correct today, but it can change out of sight of this code later. 

Please implement this algorithm: 

make tracksPositionRowIds a QHash<int, int> where key is the position and value the row id. 
than use "newPositions" to store the new position. After shuffling, iterate over newPosition[i], lookup the row id using the "position[i]" and write the new position to DB.